### PR TITLE
feat: add Listen method and optimize performance

### PR DESCRIPTION
**Pull request by devhaozi**
## 📑 Description

Closes https://github.com/goravel/goravel/issues/?

@coderabbitai summary

## ✅ Checks

- [x] Added test cases for my code

### CODE REVIEW

File: context_request.go [hunk 1]
```diff
@@ -34,12 +42,16 @@ type ContextRequest struct {
 }
 
 func NewContextRequest(ctx *Context, log log.Log, validation contractsvalidate.Validation) contractshttp.ContextRequest {
+	request := contextRequestPool.Get().(*ContextRequest)
 	httpBody, err := getHttpBody(ctx)
 	if err != nil {
-		LogFacade.Error(fmt.Sprintf("%+v", err))
+		log.Error(fmt.Sprintf("%+v", errors.Unwrap(err)))
```
Comment (hwbrzzl) [inline context_request.go:48]: `Unwrap` can't return correct error in some cases.

```suggestion
		log.Error(fmt.Sprintf("%+v", err))
```

File: context_request.go [hunk 2]
```diff
@@ -34,12 +42,16 @@ type ContextRequest struct {
...
 	}
-
-	return &ContextRequest{ctx: ctx, instance: ctx.instance, httpBody: httpBody, log: log, validation: validation}
+	request.ctx = ctx
+	request.instance = ctx.instance
+	request.httpBody = httpBody
+	request.validation = validation
```
Comment (hwbrzzl) [inline context_request.go:53]:
```suggestion
	request.log = log
	request.validation = validation
```

File: route.go [hunk 3]
```diff
@@ -128,6 +130,33 @@ func (r *Route) GlobalMiddleware(middlewares ...httpcontract.Middleware) {
...
 
+// Listen listen server
+// Listen 监听服务器
+func (r *Route) Listen(l net.Listener) error {
+	return r.instance.Listener(l)
+}
+
+// ListenTLS listen TLS server
+// ListenTLS 监听 TLS 服务器
+func (r *Route) ListenTLS(l net.Listener, certFile, keyFile string) error {
```
Comment (hwbrzzl) [inline route.go:141]: Ditto with the gin PR

File: route_test.go [hunk 4]
```diff
@@ -45,6 +46,129 @@ func TestFallback(t *testing.T) {
...
+				}
+				resp, err := http.Get(hostUrl)
+				assert.Nil(t, err)
+				defer resp.Body.Close()
+
+				body, err := io.ReadAll(resp.Body)
+				assert.Nil(t, err)
+				assert.Equal(t, "{\"Hello\":\"Goravel\"}", string(body))
+			}
+			mockConfig.AssertExpectations(t)
```
Comment (hwbrzzl) [inline route_test.go:107]: Since we are using `configmocks.NewConfig(t)`, the `AssertExpectations` method can be removed.

File: route.go [hunk 5]
```diff
@@ -128,6 +130,39 @@ func (r *Route) GlobalMiddleware(middlewares ...httpcontract.Middleware) {
...
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		Certificates: []tls.Certificate{
+			cert,
+		},
+		GetCertificate: tlsHandler.GetClientInfo,
+	}
+
+	r.instance.SetTLSHandler(tlsHandler)
+	return r.Listen(tls.NewListener(l, tlsConfig))
```
Comment (hwbrzzl) [inline route.go:163]: Do we need this part?

image

Comment (devhaozi) [reply route.go:163]: > Do we need this part?
> 
> image

No, it includes in `Listen`

Comment (hwbrzzl) [reply route.go:163]: Oh, yes

**Review by hwbrzzl — APPROVED** [review]
LGTM

**Pull request merged**